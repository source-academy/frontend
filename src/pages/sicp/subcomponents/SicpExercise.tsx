import { Button, Card, Collapse, Elevation } from '@blueprintjs/core';
import * as React from 'react';

type SicpExerciseProps = OwnProps;
type OwnProps = {
  title: string;
  body: JSX.Element;
  solution: JSX.Element | undefined;
};

export const noSolutionPlaceholder = (
  <span>
    There is currently no solution available for this exercise. This textbook adaptation is a
    community effort. Do consider contributing by providing a solution for this exercise.
    Instructions on how to contribute can be found at{' '}
    <a href="https://github.com/source-academy/sicp/wiki/Contributing-Exercise-Solutions">
      https://github.com/source-academy/sicp/wiki/Contributing-Exercise-Solutions
    </a>
    .
  </span>
);

const SicpExercise: React.FC<SicpExerciseProps> = props => {
  const [isOpen, setIsOpen] = React.useState(false);

  const onClick = () => {
    setIsOpen(!isOpen);
  };

  return (
    <Card className="sicp-exercise" interactive={false} elevation={Elevation.ONE}>
      <b>{props.title}</b>
      <div>{props.body}</div>
      <div className="sicp-button-container">
        <Button onClick={onClick} large={true} className="sicp-show-solution-button">
          {isOpen ? 'Hide Solution' : 'Show Solution'}
        </Button>
      </div>
      <Collapse className="sicp-solution" isOpen={isOpen}>
        {props.solution ? props.solution : noSolutionPlaceholder}
      </Collapse>
    </Card>
  );
};

export default SicpExercise;
