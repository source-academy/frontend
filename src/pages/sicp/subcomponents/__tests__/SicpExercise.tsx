import { Collapse } from '@blueprintjs/core';
import { mount } from 'enzyme';

import SicpExercise, { noSolutionPlaceholder } from '../SicpExercise';

describe('Sicp exercise renders', () => {
  test('correctly', () => {
    const props = {
      title: 'Title',
      body: <div>body</div>,
      solution: <div>solution</div>
    };

    const tree = mount(<SicpExercise {...props} />);
    expect(tree.debug()).toMatchSnapshot();
  });

  test('correctly with solution', () => {
    const props = {
      title: 'Title',
      body: <div>body</div>,
      solution: <div>solution</div>
    };

    const wrapper = mount(<SicpExercise {...props} />);
    wrapper.find('button').simulate('click');
    expect(wrapper.find(Collapse).text()).toEqual('solution');
  });

  test('correctly without solution', () => {
    const props = {
      title: 'Title',
      body: <div>body</div>,
      solution: undefined
    };

    const wrapper = mount(<SicpExercise {...props} />);
    wrapper.find('button').simulate('click');
    expect(wrapper.find(Collapse).text()).toEqual(mount(noSolutionPlaceholder).text());
  });
});
