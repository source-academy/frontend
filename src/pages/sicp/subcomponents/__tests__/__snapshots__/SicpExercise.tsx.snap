// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sicp exercise renders correctly 1`] = `
"<SicpExercise title=\\"Title\\" body={{...}} solution={{...}}>
  <Blueprint3.Card className=\\"sicp-exercise\\" interactive={false} elevation={1}>
    <div className=\\"bp3-card bp3-elevation-1 sicp-exercise\\">
      <b>
        Title
      </b>
      <div>
        <div>
          body
        </div>
      </div>
      <div className=\\"sicp-button-container\\">
        <Blueprint3.Button onClick={[Function: onClick]} large={true} className=\\"sicp-show-solution-button\\">
          <button type=\\"button\\" onClick={[Function: onClick]} className=\\"bp3-button bp3-large sicp-show-solution-button\\" disabled={[undefined]} onBlur={[Function (anonymous)]} onKeyDown={[Function (anonymous)]} onKeyUp={[Function (anonymous)]} tabIndex={[undefined]}>
            <Blueprint3.Icon icon={[undefined]} />
            <span className=\\"bp3-button-text\\">
              Show Solution
            </span>
            <Blueprint3.Icon icon={[undefined]} />
          </button>
        </Blueprint3.Button>
      </div>
      <Blueprint3.Collapse className=\\"sicp-solution\\" isOpen={false} component=\\"div\\" keepChildrenMounted={false} transitionDuration={200}>
        <div className=\\"bp3-collapse sicp-solution\\" style={{...}}>
          <div className=\\"bp3-collapse-body\\" style={{...}} aria-hidden={false} />
        </div>
      </Blueprint3.Collapse>
    </div>
  </Blueprint3.Card>
</SicpExercise>"
`;
